### ISSUE TITLE
Why I got different results from training exactly same Convnets every time

### ISSUE THREAD

**Issue by yushuinanrong** [vlfeat/matconvnet#598]
Hi guys,

I'm using convnets to do pose estimation and try to train the net from scracth. However, even I make the architecture and parameters same, I will get quite different results every time. My dataset contains ~30000 gray scale images adn each has size 32*32. My network has 3 convolutional layers and 2 fully connected layers. Of course, there are dropout layers and batch-normalization layers. 
The difference of the final results is not neligble. For example. the Mean-Absolut-Error of head pose can be 9.8 degree in the first training, but decease to 9.0 degree in the second training.  Anyone has the same issue? 
By the way, I'm not using Cuda-CNN. 

Any idea is greatly appreciated. 

Thanks

**Comment by bazilas** [comment]
Do you initialise the weights with the same random values every time?

**Comment by yushuinanrong** [comment]
@bazilas 
Hi Bazilas,

Thanks for your reply. I initialize my network using xavier's method. If I understand right, this initialization method should remain similar if the numbers of input and output nerons are same.

Here is one example of my initialization 
net.layers{end+1} = struct('type', 'conv', ...
'weights', {xavier(5,5,16,20)}, ...
'stride', 1, ...
'pad', 0) ;

Thanks
Jiawei

**Comment by bazilas** [comment]
Also, use the same seed for the random number generator (e.g. rng(1)  at the beginning of your code) before performing any initialisation, but don't forget to remove it when you are done with your sanity checks.

Vasilis

**Comment by yushuinanrong** [comment]
@bazilas 
Hi Vasilis,

Thanks for your suggestion. After I fixing the seed, the Convnets could converge to similar points each run. 

Jiawei

**Comment by peiyunh** [comment]
Hi @yushuinanrong , were you using GPU back then?

**Comment by yushuinanrong** [comment]
I was using GPU always. Cudnn could bring inconsistence of the final resuts
tho

**Comment by peiyunh** [comment]
My impression is calling `rng` does not fix CUDA's random seed. So if your network involves dropout or other operations involving the random number generator, you may want to find a way to fix CUDA's random seed as well.

**Comment by nightrome** [comment]
This fixes the seed for the CPU and GPU
```
randSeed = 42;
rng(randSeed);
randStream = parallel.gpu.RandStream('CombRecursive', 'Seed', randSeed);
parallel.gpu.RandStream.setGlobalStream(randStream);
```

However CuDNN is known to be non-deterministic with the default settings. If processing the same code twice, the temporal order of executions may vary and therefore the results may be different, because in general a + b ~= b + a in Computers.

**Comment by peiyunh** [comment]
@nightrome It works for me. Thanks for sharing it.

**Comment by ng0323** [comment]
Hi, where should I insert the code to fix the GPU seed as mentioned by @nightrome ?
I do not use CuDNN, would that still cause indeterminism in GPU?
Another source of randomness is in my get_batch() function. I do random cropping and flipping. But inside the get_batch() function I already do random seeding (using rng) before calling the random number function.  Still my results are not repeatable, so I'm not sure where is the problem from.

**Comment by nightrome** [comment]
Run the code before anything random happens, before you even call Matconvnet.
Parallelism might be another source of randomness, so don't use multiple GPUs or parfor loops.
Other than that I don't know. Debug your code line by line and see when the state of the random number generator changes (`sprev = rng; sprev.State`).

**Comment by saskra** [comment]
Unfortunately, the proposal by @nightrome did not help us to get reproducible results. (We did not use cuDNN.)
